accepted/2023/preview-apis/preview-apis.md

```diff
@@ -0,0 +1,279 @@
+# Experimental APIs
+
+**Owner** [Immo Landwerth](https://github.com/terrjobst)
+
+In .NET 6, we've added a way to [ship preview features][preview-features] in
+otherwise stable releases. It works by tagging APIs (and language features) as
+preview features and requiring the user's project to explicitly opt-into using
+them via the `<EnablePreviewFeatures>` property.
+
+The feature was primarily designed to accommodate platform-level preview
+features, that is for features that span runtime, library, and language, such as
+generic math. However, we also believed that the feature would work for
+libraries that want to offer preview functionality in otherwise stable NuGet
+packages.
+
+Since then, we've seen requests from partners who have tried using it and hit
+limitations that come from guardrails we put into place due to the complex
+nature of platform-level preview features.
+
+The goal of this feature is to find a design that works for both, platform-level
+preview features and library-level preview features.
+
+Let's first define these two terms:
+
+* **platform-level preview feature**. These are features that rely on preview
+  behaviors across runtime, core library, tooling, and languages. The nature of
+  these features is that they are in support of each other. For example, generic
+  math requires new runtime behaviors for interfaces, which the language has to
+  support, and of course new interfaces and methods in core library that
+  leverages them. If you want to use generic math, you need a combination of
+  runtime, framework, and language that were designed to work in conjunction.
+
+* **library-level preview feature**. These are library APIs that don't require
+  any platform-level preview features to be consumed. The only difference to
+  regular APIs is that these are in preview so their shape isn't stable and thus
```

Comment:
not just not stable, but may be entirely removed, even in a minor version.

Reply:
Being explicit that minor versions may remove api's would be really useful for CoreWCF. We have a hard commitment not to remove api's between major version releases, but our major version lifetime is a lot longer than .NET, and the time between our minor releases is relatively short. We've struggled with how to manage introducing new features which might need to have their api changed based on feedback. There might be high confidence in the code quality, but some uncertainty of the new public api's.

Reply:
Do we also want to offer any guidance about servicing? Should consumers assume that even security patches might break experimental API?
A different way of asking this is -- should consumers assume they are in an unsupported state if their app uses experimental API (because a patch might break them anytime by removing the API) or is it a reasonable configuration to deploy in production, so long as they've tested (because only a minor version update would break the API)
I'm guessing any guidance woudl be -- if you deploy to production with an experimental API, you should make no assumptions about even a security patch breaking you.

Reply:
I feel like our guidance should be as lax as possible, and we should direct libraries to make stronger guarantees if their scenario allows for it. For example, our guidance should be "Any experimental API has no shape or stability guarantees and no support guarantees", but CoreWCF could choose to guarantee "Any experimental API will not change shape in a security patch" for any such-attributed APIs in CoreWCF. This gives the most flexibility for developers to provide experimental features in their libraries without ending up back in a situation where we need to introduce a "SuperExperimental attribute" that signifies something looser than we define here.

Reply:
Sounds reasonable.

Reply:
+1 

Also may be all experimental features are not same as well. 

For example, Azure services might flight many features but may decide to follow-up different path with each feature/API in the same library.

Reply:
@danmoseley

> should consumers assume they are in an unsupported state if their app uses experimental API 

Yes. I'd argue that's the definition of using a preview/experimental/beta API. They aren't stable and thus unsupported. RCs are an exception to this rule but I think we can safely ignore that nuance.

> if you deploy to production with an experimental API, you should make no assumptions about even a security patch breaking you.

Agreed.

@jkoritzinsky 

> I feel like our guidance should be as lax as possible

I think you mean the opposite: we want the most amount of flexibility, so I think the guidance should be as definitive as possible, along the lines of what Dan was suggesting.

Reply:
Whatever the right term is, I think the guidance should allow the most flexibility possible.

Reply:
I think the semantics need to be clear, and keep the guidance to a minimum:


APIs marked as experimental indicate that the author of the API is not committing to supporting this API over time, and the quality of the API's implementation may be lower than the rest of the API surface. If you take a dependency on an experimental API, you may find that the API signature changes or the API is completely removed the next time you upgrade to a new version of the assembly containing the API.

And then the Experimental attribute should have a note property that lets the author provide more detailed guidance if needed. "We're pretty confident about this API, we expect to remove this attribute in the 1.2 release", "Please don't use this right now, we're still in active development". Stuff like that.

Reply:
Can we embed my wording suggestion into the doc? I think it hits all the salient points.